tests/baselines/reference/arrowFunctionReturnTypeErrorSpan.errors.txt

```diff
@@ -0,0 +1,28 @@
+arrowFunctionReturnTypeErrorSpan.ts(3,3): error TS2322: Type 'string' is not assignable to type 'number'.
+arrowFunctionReturnTypeErrorSpan.ts(12,25): error TS2304: Cannot find name 'missing'.
+arrowFunctionReturnTypeErrorSpan.ts(7,15): error TS2322: Type 'string' is not assignable to type 'number'.
+arrowFunctionReturnTypeErrorSpan.ts(10,15): error TS2322: Type 'string' is not assignable to type 'number'.
+
+
+==== arrowFunctionReturnTypeErrorSpan.ts (4 errors) ====
+    // block body
+    const a = (): number => {
+      return "foo";
+      ~~~~~~
+!!! error TS2322: Type 'string' is not assignable to type 'number'.
+    };
+    
+    // expression body
+    const b = (): number => "foo";
+                  ~~~~~~
+!!! error TS2322: Type 'string' is not assignable to type 'number'.
+    
+    type F<T> = T;
+    const c = (): F<number> => "foo";
+                  ~
+!!! error TS2322: Type 'string' is not assignable to type 'number'.
+    
+    const d = (): number => missing;
+                            ~~~~~~~
+!!! error TS2304: Cannot find name 'missing'.

```

Comment:
My change broke this test. I only want to move the span for errors relating to a return type mismatch. How can I do that?

Reply:
I'm not sure it is possible to disambiguate these errors in `getErrorSpanForNode` as both return/signature mismatch and errors in the root expression use the same node.

I think what needs to be modified is the node being targeted when a signature mismatch happens.

I believe @Andarist gave good pointers in this comment https://github.com/microsoft/TypeScript/issues/57866#issuecomment-2225630086

I got this to work:

```diff
diff --git a/src/compiler/checker.ts b/src/compiler/checker.ts
index 04facba319..16b2b5f65a 100644
--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -45798,7 +45798,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             : exprType;
 
         const effectiveExpr = expr && getEffectiveCheckNode(expr); // The effective expression for diagnostics purposes.
-        const errorNode = inReturnStatement && !inConditionalExpression ? node : effectiveExpr;
+        const parent = node.parent;
+        const errorNode =
+          inConditionalExpression ? effectiveExpr :
+          inReturnStatement ? node :
+          isArrowFunction(parent) && parent.type !== undefined ? parent.type:
+          effectiveExpr;
 
         // If the return type is not narrowable, we simply check if the return expression type is assignable to the return type.
         if (!(unwrappedReturnType.flags & (TypeFlags.IndexedAccess | TypeFlags.Conditional)) || !couldContainTypeVariables(unwrappedReturnType)) {

```

with the following changes (I believe most of them make sense 🤔, but some might be confusing):

```diff
--- tests/baselines/reference/mappedTypeConstraints2.errors.txt	2024-12-17 19:48:25.745594826 -0500
+++ tests/baselines/local/mappedTypeConstraints2.errors.txt	2024-12-17 19:52:57.234587430 -0500
@@ -6,7 +6,7 @@
 mappedTypeConstraints2.ts(42,7): error TS2322: Type 'Mapped6<K>[keyof Mapped6<K>]' is not assignable to type '`_${string}`'.
   Type 'Mapped6<K>[string] | Mapped6<K>[number] | Mapped6<K>[symbol]' is not assignable to type '`_${string}`'.
     Type 'Mapped6<K>[string]' is not assignable to type '`_${string}`'.
-mappedTypeConstraints2.ts(51,57): error TS2322: Type 'Foo<T>[`get${T}`]' is not assignable to type 'T'.
+mappedTypeConstraints2.ts(51,52): error TS2322: Type 'Foo<T>[`get${T}`]' is not assignable to type 'T'.
   'T' could be instantiated with an arbitrary type which could be unrelated to 'Foo<T>[`get${T}`]'.
 mappedTypeConstraints2.ts(82,9): error TS2322: Type 'ObjectWithUnderscoredKeys<K>[`_${K}`]' is not assignable to type 'true'.
   Type 'ObjectWithUnderscoredKeys<K>[`_${string}`]' is not assignable to type 'true'.
@@ -75,7 +75,7 @@
     };
     
     const get = <T extends string>(t: T, foo: Foo<T>): T => foo[`get${t}`];  // Type 'Foo<T>[`get${T}`]' is not assignable to type 'T'
-                                                            ~~~~~~~~~~~~~~
+                                                       ~
 !!! error TS2322: Type 'Foo<T>[`get${T}`]' is not assignable to type 'T'.
 !!! error TS2322:   'T' could be instantiated with an arbitrary type which could be unrelated to 'Foo<T>[`get${T}`]'.
     
```

```diff
--- tests/baselines/reference/conditionalTypes1.errors.txt	2024-12-17 19:48:25.744454937 -0500
+++ tests/baselines/local/conditionalTypes1.errors.txt	2024-12-17 19:53:07.197814413 -0500
@@ -73,7 +73,7 @@
   Type 'string | number' is not assignable to type 'ZeroOf<T>'.
     Type 'string' is not assignable to type 'ZeroOf<T>'.
 conditionalTypes1.ts(263,9): error TS2403: Subsequent variable declarations must have the same type.  Variable 'z' must be of type 'T1', but here has type 'Foo<T & U>'.
-conditionalTypes1.ts(288,43): error TS2322: Type 'T95<U>' is not assignable to type 'T94<U>'.
+conditionalTypes1.ts(288,33): error TS2322: Type 'T95<U>' is not assignable to type 'T94<U>'.
   Type 'number | boolean' is not assignable to type 'T94<U>'.
     Type 'number' is not assignable to type 'T94<U>'.
       Type 'boolean' is not assignable to type 'true'.
@@ -465,7 +465,7 @@
     type T95<T> = T extends string ? boolean : number;
     const f44 = <U>(value: T94<U>): T95<U> => value;
     const f45 = <U>(value: T95<U>): T94<U> => value;  // Error
-                                              ~~~~~
+                                    ~~~~~~
 !!! error TS2322: Type 'T95<U>' is not assignable to type 'T94<U>'.
 !!! error TS2322:   Type 'number | boolean' is not assignable to type 'T94<U>'.
 !!! error TS2322:     Type 'number' is not assignable to type 'T94<U>'.

```

```diff
--- tests/baselines/reference/typeSatisfaction_errorLocations1.errors.txt	2024-12-17 17:21:06.684977990 -0500
+++ tests/baselines/local/typeSatisfaction_errorLocations1.errors.txt	2024-12-17 19:55:16.781963398 -0500
@@ -26,13 +26,13 @@
 typeSatisfaction_errorLocations1.ts(39,3): error TS2322: Type 'string' is not assignable to type 'number'.
 typeSatisfaction_errorLocations1.ts(43,3): error TS2322: Type 'string' is not assignable to type 'number'.
 typeSatisfaction_errorLocations1.ts(43,16): error TS1360: Type 'string' does not satisfy the expected type 'number'.
-typeSatisfaction_errorLocations1.ts(46,22): error TS2741: Property 'a' is missing in type '{}' but required in type '{ a: true; }'.
+typeSatisfaction_errorLocations1.ts(46,6): error TS2741: Property 'a' is missing in type '{}' but required in type '{ a: true; }'.
 typeSatisfaction_errorLocations1.ts(47,24): error TS2322: Type 'number' is not assignable to type 'true'.
-typeSatisfaction_errorLocations1.ts(48,21): error TS2322: Type '{ a: number; }' is not assignable to type '{ a: true; }'.
+typeSatisfaction_errorLocations1.ts(48,6): error TS2322: Type '{ a: number; }' is not assignable to type '{ a: true; }'.
   Types of property 'a' are incompatible.
     Type 'number' is not assignable to type 'true'.
-typeSatisfaction_errorLocations1.ts(50,23): error TS2741: Property 'a' is missing in type '{}' but required in type '{ a: true; }'.
-typeSatisfaction_errorLocations1.ts(51,24): error TS2741: Property 'a' is missing in type '{}' but required in type '{ a: true; }'.
+typeSatisfaction_errorLocations1.ts(50,6): error TS2741: Property 'a' is missing in type '{}' but required in type '{ a: true; }'.
+typeSatisfaction_errorLocations1.ts(51,6): error TS2741: Property 'a' is missing in type '{}' but required in type '{ a: true; }'.
 
 
 ==== typeSatisfaction_errorLocations1.ts (24 errors) ====
@@ -133,7 +133,7 @@
     }
     
     ((): { a: true } => ({}) satisfies unknown)();
-                         ~~
+         ~~~~~~~~~~~
 !!! error TS2741: Property 'a' is missing in type '{}' but required in type '{ a: true; }'.
 !!! related TS2728 typeSatisfaction_errorLocations1.ts:46:8: 'a' is declared here.
     ((): { a: true } => ({ a: 1 }) satisfies unknown)();
@@ -141,17 +141,17 @@
 !!! error TS2322: Type 'number' is not assignable to type 'true'.
 !!! related TS6500 typeSatisfaction_errorLocations1.ts:47:8: The expected type comes from property 'a' which is declared here on type '{ a: true; }'
     ((): { a: true } => obj1 satisfies unknown)();
-                        ~~~~
+         ~~~~~~~~~~~
 !!! error TS2322: Type '{ a: number; }' is not assignable to type '{ a: true; }'.
 !!! error TS2322:   Types of property 'a' are incompatible.
 !!! error TS2322:     Type 'number' is not assignable to type 'true'.
     
     ((): { a: true } => (({}) satisfies unknown) satisfies unknown)();
-                          ~~
+         ~~~~~~~~~~~
 !!! error TS2741: Property 'a' is missing in type '{}' but required in type '{ a: true; }'.
 !!! related TS2728 typeSatisfaction_errorLocations1.ts:50:8: 'a' is declared here.
     ((): { a: true } => ((({}) satisfies unknown)) satisfies unknown)();
-                           ~~
+         ~~~~~~~~~~~
 !!! error TS2741: Property 'a' is missing in type '{}' but required in type '{ a: true; }'.
 !!! related TS2728 typeSatisfaction_errorLocations1.ts:51:8: 'a' is declared here.
     
\ No newline at end of file

```
and

```diff
--- tests/baselines/reference/arrowFunctionReturnTypeErrorSpan.errors.txt	2024-12-17 19:48:25.742635790 -0500
+++ tests/baselines/local/arrowFunctionReturnTypeErrorSpan.errors.txt	2024-12-17 19:55:36.318525677 -0500
@@ -1,6 +1,6 @@
 arrowFunctionReturnTypeErrorSpan.ts(3,3): error TS2322: Type 'string' is not assignable to type 'number'.
-arrowFunctionReturnTypeErrorSpan.ts(7,25): error TS2322: Type 'string' is not assignable to type 'number'.
-arrowFunctionReturnTypeErrorSpan.ts(10,28): error TS2322: Type 'string' is not assignable to type 'number'.
+arrowFunctionReturnTypeErrorSpan.ts(7,15): error TS2322: Type 'string' is not assignable to type 'number'.
+arrowFunctionReturnTypeErrorSpan.ts(10,15): error TS2322: Type 'string' is not assignable to type 'number'.
 arrowFunctionReturnTypeErrorSpan.ts(12,25): error TS2304: Cannot find name 'missing'.
 
 
@@ -14,12 +14,12 @@
     
     // expression body
     const b = (): number => "foo";
-                            ~~~~~
+                  ~~~~~~
 !!! error TS2322: Type 'string' is not assignable to type 'number'.
     
     type F<T> = T;
     const c = (): F<number> => "foo";
-                               ~~~~~
+                  ~~~~~~~~~
 !!! error TS2322: Type 'string' is not assignable to type 'number'.
     
     const d = (): number => missing;

```

Reply:
> I think what needs to be modified is the node being targeted when a signature mismatch happens.

this sounds about right 👍